### PR TITLE
a set of complicated optimizations

### PR DESCRIPTION
**Pull request by se-ti**
No silver bullet, but a bunch of techniques.

- be cache friendly
- be processor branch predictor friendly
- less memory allocations
- don't compute unnecessary
- use effective functions
- something else?

### Be cache friendly
Processors do not like cache misses. Try to organize data in proper way.
During rendering we need them to go row by row -- let's fill them such a way: outer cycles go by rows (y), inner cycles -- by columns (x).

### Less memory allocations
Object-based value allocated memory regularly (if (data[pos]) ... else data[pos] = v;). Computing large heatmap turned out to be not really a computing, but a memory allocation (and garbage collection!).
Replace object with once-allocated array.

### Be branch predictor friendly
Try to use as few conditional operators as you can.
- check borders once, and use highly optimized function, that doesn't check borders. Leave complicated checks for circles that cross image border.
- allocate all memory once, fill it with zeroes, get rid of  if (data[pos]) data[pos] += v; else data[pos] = v;
- use array, not an object for values. It's less memory effective for "empty" heatmaps, especially on large screens, but it speeds up well-filled heatmaps dramatically.
- compute loop ranges as precise, as possible.

### Don't compute unnecessary
- don't compute if (dx^2 + dy ^2 < radius ^2). Compute correct range for each scanline. ~1/4 inner iterations less.
- circles are supersymmetric, compute weight once, use it for 8 points! -- 8 times less Math.pow(dist, degree) computations!
- don't cross borders :)
- compute main offsets, alter them with ++ or +=; Sorry, less semantic, but more efficient than data[x + y * params.width] :(

### Use effective functions
Math.min, and Math.max are more semantic, but if (x < min) min = x and ternary operator are much, much more effective.
Replace "for in" with "for" in rendering. ~400 ms -> 60-80 ms. Just a price, i pay for serializing / deserializing heatmap data to communicate  with worker.

Something, i forgot?

All these and previous optimizations allowed me to speed my heatmaps (15K points, with high peaks) up by 10-12 times.
Now it takes ~1 second including data transfer to and from worker.

### CODE REVIEW

**Comment by se-ti** [comment]
Topics to research.
- think on using well-seasoned Bresenham algorithm for circle border instead of Math.floor(Math.sqrt(radiusSq - dySq));
- circle weights don't depend on source data, just on radius. So we can precompute them once for the largest radius (not memoize -- see branch predictor friendliness)! Classical time-memory tradeoff, but it can be expensive and ineffective for some data and screen sizes.
- in my cases webworker communication expenses were equal or close to worker's work time (300-400 ms). That could be the reason to compute heatmap in UI thread. Surprisingly, it could be more responsive! 
Probably there should be an estimate (incl. screen size, max value, etc.) when to do it on ui thread, or on worker's one.

And the last, but not least. I am not sure, but i won't be too surprised, if some who knows math well would say, that all we do is just a convolution of the source data with some kernel.
AFAIK, an efficient way to compute convolution, is to compute FFT for source data and kernel, multiply their spectrums, and compute inverse FFT.
FFT on js... sounds mad ^)

**Review by sunng87 — APPROVED** [review]

File: src/heatcanvas-worker.js [hunk 1]
```diff
@@ -24,47 +24,159 @@ onmessage = function(e){
...
+        value[base] += data;
     }
     postMessage({'value': value, 'width': params.width, 'height': params.height});
 }
+
+/* Uses fact that circle on the screen has 4 axes of symmetry: x = 0, y = 0, y = x, y = -x,
+ so it computes values for ~1/8 of points and copies them to symmetrical ones
+
+ Does not check bounds of the image!
+ */
```
Comment (sunng87) [inline src/heatcanvas-worker.js:126]: :+1:

**Comment by sunng87** [comment]
Thank you very much! The `fastCalc` was impressive. 

By the way, do you think we can benefit from technology like webassembly on this compute heavy task?

**Comment by sunng87** [comment]
Just added you write permission for this repo. Feel free to change anything if I'm late in response.

**Comment by se-ti** [comment]
No problem! I prefer people who read code to those, who merge fast :)
And thanks for access!

I have no experience with web assembly, and no idea, whether it would be useful. 
Are there any highload use cases?
Are there any people who feed this canvas with 10K points? Who want to feed it with 100K points?

Right now we loose a lot of time on UI <-> web worker communication.  And... my today's experiments show, that migrating from generic Array to Float32Array almost eliminate these losses.
~140 ms just to return data to ui thread shrink to 4-5 ms for 1300x700px canvas. 

Applying typed array to pass source data to web worker seems to be very useful.

**Comment by sunng87** [comment]
That's interesting. Do you know why the typed array works that faster? Does it apply to all browser implementation?

**Comment by se-ti** [comment]
I have tested it in actual versions of desktop Firefox and Chrome. Both showed similar results.

I am not sure, but MDN and some remnants in the event object hint that inner mechanism of worker communication is a kind of pipe. It's much easier to send data as a BLOB (Hey! 900k of floats coming! memcpy), than a generic array, i suppose.

But it's strange, that sending typed arrays to worker was of no effect :(
It still takes 20-25 ms (Chrome) or 40-60ms (FF) to send data and 1-2-5 ms to return much more data back.

P.S. I thought on precomputing weights. Allocating 1/4 of the biggest circles (but no more than canvas size ) for cache seems to be reasonable price even for 4K fullscreen canvas. It will be the next step, if you don't mind.

P.P.S. I realized that migrating from object to arrays should have broken custom background color :(  
It's possible to fix it for default palette, but i don't see efficient fix for an arbitrary colorscheme.
color = value > 0 ? color : bkColor doesn't look efficient enough :(

**Comment by se-ti** [comment]
Seems that delays when sending data to worker were inspired by LeafLet, that tried to move and zoom layer, to fit data to viewport, and show heatmap layer almost simultaneously.
Simple turning heatmap on and off showed reasonable latencies ~3-5 ms.

**Pull request merged**